### PR TITLE
Use PSA EC-JPAKE in TLS (1.2)

### LINKED ISSUE #5847: EC-JPAKE 2: use in TLS (1.2)
TLS 1.2 has optional support for a key exchange based on EC J-PAKE, defined as part of the [Thread standard](https://www.threadgroup.org/ThreadSpec), see `MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED`. Currently the implementation always uses the `mbedtls_ecjpake` API regardless of whether `MBEDTLS_USE_PSA_CRYPTO` is enabled.

This task it to make it use the PSA API when `MBEDTLS_USE_PSA_CRYPTO` is enabled.

Depends on:
- #4745 and #4153 for the PSA EC J-PAKE API to be both stabilized and implemented.
- #5978 for the ad-hoc KDF that TLS 1.2 needs to get its PMS from the generic shared secret of PSA.

### PR DESCRIPTION
**Pull request by superna9999**
## Description
TLS 1.2 has optional support for a key exchange based on EC J-PAKE, defined as part of the [Thread standard](https://www.threadgroup.org/ThreadSpec), see MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED. Currently the implementation always uses the mbedtls_ecjpake API regardless of whether MBEDTLS_USE_PSA_CRYPTO is enabled.

This task it to make it use the PSA API when MBEDTLS_USE_PSA_CRYPTO is enabled.

Resolves #5847 

## Status
**IN DEVELOPMENT**

## Requires Backporting
NO 

## Migrations
NO

## Additional comments
Depends on #6115 to be merged

## Todos
- [ ] Tests

## Steps to test or reproduce
ssl-opt.sh must run clean

### CODE REVIEW

**Comment by mpg** [comment]
Next time you update this, can you also remove the line about EC J-PAKE in `docs/use-psa-crypto.md`?

**Comment by mpg** [comment]
Note: the failure in ABI-API checking is not a problem, see https://github.com/Mbed-TLS/mbedtls/pull/6115#issuecomment-1243417738

**Comment by mpg** [comment]
Now that #6115 has been merged, this should be rebased. @superna9999 Can you take care of this? Otherwise please let us know and we'll handle it.

**Comment by superna9999** [comment]
@mpg done!

**Comment by mpg** [comment]
Great, thanks!

File: library/ssl_tls12_client.c [hunk 1]
```diff
@@ -3244,6 +3387,35 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
...
+                psa_destroy_key( ssl->handshake->psa_pake_password );
+                psa_pake_abort( &ssl->handshake->psa_pake_ctx );
+                MBEDTLS_SSL_DEBUG_RET( 1 , "psa_pake_output", status );
+                return( psa_ssl_status_to_mbedtls( status ) );
+            }
+
+            output_offset += output_len;
+        }
+
+        content_len = output_offset;
```
Comment (mpg) [inline library/ssl_tls12_client.c:3417]: Question: why do we stop there instead of also computing the PMS as we do in the non-PSA case? If there's a reason, it's not obvious to me.

Generally speaking, it's a pre-existing issue with the code that the PMS is computed in various places for various key exchanges and this makes things really hard to follow at times. Let's not introduce more variations unless we have a good reason.

**Review by mpg — CHANGES_REQUESTED** [review]
I only found very minor things by reviewing, but did some more testing and the results were not good, see my next message.

**Comment by mpg** [comment]
It's hard to validate correctness (are we using the correct format? computing the expected value for the shared secret? etc), not just for this PR, but for the entire EC J-PAKE line of work, because the key exchange is randomized and we can't use known-answer tests.

So, I guess the only thing we can do is test interop with other implementations. I'm not aware of other implementations of the ECJPAKE key exchange in TLS out there, but now we have two in Mbed TLS: the version that uses PSA and the one that doesn't. Let's see if they can talk to each other:
```sh
git checkout -- include/mbedtls/mbedtls_config.h
scripts/config.py set MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED
make clean

make -C programs ssl/ssl_server2 ssl/ssl_client2
cp programs/ssl/ssl_server2 s2_no_use_psa
cp programs/ssl/ssl_client2 c2_no_use_psa


scripts/config.py set MBEDTLS_USE_PSA_CRYPTO
make clean

make -C programs ssl/ssl_server2 ssl/ssl_client2
make -C programs test/udp_proxy test/query_compile_time_config

P_SRV=../s2_no_use_psa tests/ssl-opt.sh -f ECJPAKE
P_CLI=../c2_no_use_psa tests/ssl-opt.sh -f ECJPAKE
```

And unfortunately the results are not looking good on my machine:
```
ECJPAKE: client not configured ......................................... PASS
ECJPAKE: server not configured ......................................... PASS
ECJPAKE: working, TLS .................................................. FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-3.log
ECJPAKE: password mismatch, TLS ........................................ FAIL
  ! pattern 'SSL - Verification of the message MAC failed' MUST be present in the Server output
  ! outputs saved to o-XXX-4.log
ECJPAKE: working, DTLS ................................................. FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-5.log
ECJPAKE: working, DTLS, no cookie ...................................... FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-6.log
ECJPAKE: password mismatch, DTLS ....................................... FAIL
  ! pattern 're-using cached ecjpake parameters' MUST be present in the Client output
  ! outputs saved to o-XXX-7.log
ECJPAKE: working, DTLS, nolog .......................................... FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-8.log
------------------------------------------------------------------------
FAILED (2 / 8 tests (0 skipped))
```

I don't have time to debug now, I can have a look next week until you beat be to it, @superna9999 

Also, based on this, I'm going to advocate that we add an `all.sh` component doing something similar to the above.

File: library/ssl_tls12_client.c (hunk 1 above)
Comment (superna9999) [reply library/ssl_tls12_client.c:3417]: Hmm it wasn't necessary before the addition of the custom SHA256 PMS for PAKE, so indeed I may have forgotten a step while migrating and it may explain why it doesn't work with the non-PSA implementation.

**Comment by superna9999** [comment]
@mpg thanks, I think I may missed something when adding support for #6115, in the first implementation it worked while developement when I implemented only the client and left the server using non-PSA. But when migrated to the custom KDF I may have missed step.

I'll add the all.sh component to this PR as you suggest.

**Comment by mpg** [comment]
Note: this will need rework once the I/O format of `psa_pake_input/output()` has been fixed to match the spec, as done in https://github.com/Mbed-TLS/mbedtls/pull/6390

**Comment by superna9999** [comment]
Finally I'm lacking time to review and update those PRs, I'm OK if those are taken over by @mprse @AndrzejKurek or @valeriosetti

**Comment by valeriosetti** [comment]
> Finally I'm lacking time to review and update those PRs, I'm OK if those are taken over by @mprse @AndrzejKurek or @valeriosetti

I'm currently trying to fix test issues reported by @mpg by rebasing this PR on top of PR #6390. Once I'll have something ready I'll open a new PR to continue this work

**Comment by superna9999** [comment]
Taken over by @valeriosetti in #6533
